test/grpc/route_guide/route-chat.cc

```diff
@@ -0,0 +1,130 @@
+      const std::string& db)
+    : RouteGuideBaseClient(target, credentials, std::move(pool), db) {}
+
+  auto RouteChat() {
+    return client_.Call<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Then(Let([&](auto& call) {
+             return DoAll(
+                        Foreach(
+                            Iterate(
+                                {MakeRouteNote(data_[0]),
+                                 MakeRouteNote(data_[1]),
+                                 MakeRouteNote(data_[2]),
+                                 MakeRouteNote(data_[3])}),
+                            [&](RouteNote note) {
+                              return call.Writer().Write(note);
+                            })
+                            | call.WritesDone(),
+                        Foreach(
+                            call.Reader().Read(),
+                            [this](RouteNote&& note) {
+                              EXPECT_EQ(
+                                  note.message(),
+                                  std::get<0>(data_[current_]) + " bounced");
+                              EXPECT_EQ(
+                                  note.location().latitude(),
+                                  std::get<1>(data_[current_]));
+                              EXPECT_EQ(
+                                  note.location().longitude(),
+                                  std::get<2>(data_[current_++]));
+                            }))
+                 | Finally([&](auto) {
+                      return call.Finish();
+                    })
+                 | Then([this](Status&& status) {
+                      CHECK(status.ok()) << status.error_code()
+                                         << ": " << status.error_message();
+                      CHECK_EQ(current_, data_.size());
+
+                      return true;
+                    });
+           }));
+  }
+
+ private:
+  size_t current_ = 0;
+  const std::vector<std::tuple<std::string, long, long>> data_ = {
+      {"First message", 0, 0},
+      {"Second message", 0, 1},
+      {"Third message", 1, 0},
+      {"Fourth message", 0, 0}};
+};
+
+TEST(RouteGuideTest, RouteChatTest) {
+  std::string db_path("test/grpc/route_guide/route_guide_db.json");
+  std::string db;
+  routeguide::GetDbFileContent(db_path, db);
+
+  RouteGuideImpl service(db);
+
+  ServerBuilder builder;
+  std::string server_address("localhost:45554");
+
+  builder.AddListeningPort(
+      server_address,
+      grpc::InsecureServerCredentials());
+
+  builder.RegisterService(&service);
+
+  auto build = builder.BuildAndStart();
+
+  ASSERT_TRUE(build.status.ok()) << build.status;
+
+  auto server = std::move(build.server);
+
+  ASSERT_TRUE(server);
+
+  Borrowable<CompletionPool> pool;
+
+  RouteGuideRouteChatClient guide(
+      server_address,
+      grpc::InsecureChannelCredentials(),
+      pool.Borrow(),
+      db);
```

Comment:
Let's create a test fixture that does all of this for us and all we need to do is call the function, e.g., `RouteChat()`.